Avoid starting a new TCL interpreter for each scenario

ahhbristow opened the pull request:
I've added support to pass in an environment variable called NEW_INTERPRETER when running Cucumber TCL.  This changes Cucumber TCL to not start a new TCL interpreter for each scenario, which drastically improves the speed of our tests by not having to source env.tcl each time (which may do a lot of initialisation work).  By passing NEW_INTERPRETER=1 into the Cucumber command, you can force a new interpreter.

This may be useful if you believe there is data leakage between scenarios due to TCL packages loaded as part of the test run retaining global or namespace variables, which could be a problem for poorly understood legacy code.

NEW_INTERPRETER defaults to 0.

ahhbristow commented:
I branched off from v0.0.6 rather than HEAD (which is unreleased and untested at the min), hence the conflict.

aslakhellesoy commented:
Could you document this in the README please?

Would it be possible to add a test for this as well?

Thanks!

Aslak

ahhbristow commented:
Will do, I'm not sure how the tests work yet as this is my first submission and haven't run them :). Just wanted to get this on a branch to see if I've taken the correct approach.

mattwynne commented:
@ahhbristow check https://github.com/cucumber/cucumber-ruby-tcl/blob/master/CONTRIBUTING.md for how to run the tests.

mattwynne commented:
Looks like a reasonable approach to me.

mattwynne commented:
It will be a good challenge to see if you can express this in a feature file.

ahhbristow commented:
Thanks @mattwynne, it took me a while to figure out how everything fits together but I've updated reset_state.feature with some new scenarios now.  I've also fixed a bug related to file sourcing where only one step definition file was being loaded regardless of how many there were, and updated one of the scenarios to capture this.

Updated the README.md as well.

mattwynne commented on lib/cucumber/tcl.rb:
```diff
@@ -13,9 +13,27 @@ module Cucumber
+          new_interpreter = 1
+      else 
+          new_interpreter = ENV['NEW_INTERPRETER'].to_i
+      end
```
you could just say:

`new_interpreter = (ENV['NEW_INTERPRETER'] || 1).to_i`

mattwynne replied:
Also, it's more idiomatic to use a boolean in Ruby, so it would be more like this:

`new_interpreter = (ENV['NEW_INTERPRETER'] == '1')`

Though that would change the behaviour and make the default be to *not* use a new interpreter.

Maybe it would be more clear to name the env var in a positive way, like `SHARE_FRAMEWORK`?

mattwynne replied:
Otherwise you could negate it, like this:

new_interpreter = (ENV['NEW_INTERPRETER'] != '0')

ahhbristow commented:
@mattwynne I've incorporated your suggestions, and added myself as a release manager in CONTRIBUTING.md.  I'll get Jowers to grant release privilege so I can put a new version out.  Thanks for reviewing my changes.

mattwynne commented:
👍

Typed with my Thumbs.

> On 19 Apr 2017, at 01:50, Shaun Bristow <notifications@github.com> wrote:
> 
> @mattwynne I've incorporated your suggestions, and added myself as a release manager in CONTRIBUTING.md. I'll get Jowers to grant release privilege so I can put a new version out. Thanks for reviewing my changes.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
> 

-- 



The Cucumber logo is the intellectual property of Cucumber Ltd, a limited 
company registered in Scotland, number 456793.

UK Headquarters: Cucumber Ltd, Drumsyniebeg, Lochgoilhead, Cairndow, 
Argyll, PA24 8AN UK.

CONFIDENTIALITY NOTICE: The information in this e-mail is confidential and 
privileged; it is intended for use solely by the individual or entity named 
as the recipient hereof. Disclosure, copying, distribution, or use of the 
contents of this e-mail by persons other than the intended recipient is 
strictly prohibited and may violate applicable laws. If you have received 
this e-mail in error, please delete the original message and notify us by 
email immediately. Thank you. Cucumber Ltd.
